crates/data-structures/src/error_stream.rs

```diff
@@ -0,0 +1,248 @@
+#[derive(thiserror::Error, Debug, Clone, Default, PartialEq, Eq)]
+pub struct ErrorStream<E>(smallvec::SmallVec<[E; 1]>);
+
+impl<E> ErrorStream<E> {
+    /// Returns an iterator over the errors in the stream.
+    pub fn iter(&self) -> impl Iterator<Item = &E> {
+        self.0.iter()
+    }
+
+    /// Returns a mutable iterator over the errors in the stream.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut E> {
+        self.0.iter_mut()
+    }
+
+    /// Returns an iterator over the errors in the stream, consuming the stream.
+    pub fn drain(&mut self) -> impl Iterator<Item = E> + '_ {
+        self.0.drain(..)
+    }
+
+    /// Unpack an error into `self`, returning `None` if there was an error.
+    /// This is not exposed because `CombineErrors` is more convenient.
+    #[inline(never)] // don't optimize this too much
+    #[cold]
+    fn unpack<T, ES: Into<ErrorStream<E>>>(&mut self, result: Result<T, ES>) -> Option<T> {
+        match result {
+            Ok(value) => Some(value),
+            Err(error) => {
+                self.0.extend(error.into().0);
+                None
+            }
+        }
+    }
+}
+
+impl<E> AddAssign<E> for ErrorStream<E> {
+    fn add_assign(&mut self, error: E) {
+        self.0.push(error);
+    }
+}
+impl<E> AddAssign<ErrorStream<E>> for ErrorStream<E> {
+    fn add_assign(&mut self, other: ErrorStream<E>) {
+        self.0.extend(other.0);
+    }
+}
+
+impl<E> From<E> for ErrorStream<E> {
+    fn from(error: E) -> Self {
+        Self(smallvec::smallvec![error])
+    }
+}
+
+/// A trait for converting a tuple of `Result`s into a `Result` of a tuple or `ErrorStream`.
+pub trait CombineErrors {
+    /// The type of the output if all results are `Ok`.
+    type Ok;
+    /// The type of the output if any result is `Err`.
+    type Error;
+
+    /// Combine errors from multiple places into one.
+    /// This can be thought of as a kind of parallel `?`.
+    ///
+    /// If your goal is to show the user as many errors as possible, you should
+    /// call this as late as possible, on as wide a tuple as you can.
+    ///
+    /// Example usage:
+    ///
+    /// ```
+    /// use spacetimedb_data_structures::error_stream::{ErrorStream, CombineErrors};
+    ///
+    /// struct MyError { cause: String };
+    ///
+    /// fn age() -> Result<i32, ErrorStream<MyError>> {
+    ///     //...
+    /// # Ok(1)
+    /// }
+    ///
+    /// fn name() -> Result<String, ErrorStream<MyError>> {
+    ///     // ...
+    /// # Ok("hi".into())
+    /// }
+    ///
+    /// fn description() -> Result<String, ErrorStream<MyError>> {
+    ///     let (age, name) = (age(), name()).combine_errors()?;
+    ///     Ok(format!("{} is {} years old", name, age))
+    /// }
+    /// ```
+    fn combine_errors(self) -> Result<Self::Ok, ErrorStream<Self::Error>>;
+}
+
+macro_rules! tuple_combine_errors {
+    ($($T:ident),*) => {
+        impl<$($T,)* E> CombineErrors for ($(Result<$T, ErrorStream<E>>,)*) {
```

Comment:
Why is this impl'd for `Result<T, ErrorStream<E>>` rather than for `Result<T, E>`? Is it possible to do some type magic to permit either `E` or `ErrorStream<E>` as the err variant of the result in this impl? Something like,

```rust
trait ExtendErrorStream<E> {
  fn add_error(&mut self, e: E);
}
impl<E> ExtendErrorStream<E> for ErrorStream<E> {
  fn add_error(&mut self, e: E) { self.0.push(e); }
}
impl<E> ExtendErrorStream<ErrorStream<E>> for ErrorStream<E> {
  fn add_error(&mut self, e: ErrorStream<E>) { self.0.extend(e.0); }
}

// macro needs both `T` ok types and `En` err types, and expands:
impl<$($T,)* $($En,)* E> CombineErrors for ($(Result<$T, $En>,)*)
where
  $(ErrorStream<E>: ExtendErrorStream<$En>,)*
{
  type Ok = ($($T,)*);
  type Error = E;
  fn combine_errors(self) -> Result<Self::Ok, ErrorStream<Self::Error>> {
    // pretty much the same, but with `add_error` instead of `unpack`.
  }
}
```

Reply:
I tried to impl for both `Result<T, E>` and `Result<T, ErrorStream<E>>`, but that fails because it overlaps. Lemme try that and see if it works.

Reply:
Ah, you still run into issues. `(Result<A, ErrorStream<E>>, Result<B, ErrorStream<E>>)` ends up ambiguous. Should the output type be `Result<(A, B). ErrorStream<E>>` or `Result<(A, B), ErrorStream<ErrorStream<E>>>`?

Reply:
Darn! Oh well.